File: problems/1001-1100/1067_digit-occurrences-in-a-range/solution_digit_dp.java
import java.util.*;

class Solution {

    // Prefix-count reduction: occurrences in [low, high] = f(high) - f(low-1).
    public int countDigitOccurrences(int d, int low, int high) {
        return (int) (countUpTo(d, high) - countUpTo(d, low - 1));
    }

    private long countUpTo(int d, long n) {
        if (n <= 0) {
            return 0;
        }
        String s = Long.toString(n);
        int length = s.length();
        int[] digits = new int[length];
        for (int i = 0; i < length; i++) {
            digits[i] = s.charAt(i) - '0';
        }
        // Free (non-tight) suffixes recur, so they are memoized per
        // (position, started): {completions, occurrences} pairs.
        long[][][] memo = new long[length][2][];
        // The all-zero completion is the number 0 and carries no
        // appearances, so the walk tallies exactly the integers 1..n.
        return solve(digits, d, 0, true, false, memo)[1];
    }

    // Each state reports how many suffix completions it admits and how many
    // appearances of d those completions contain.
    private long[] solve(int[] digits, int d, int pos, boolean tight, boolean started, long[][][] memo) {
        if (pos == digits.length) {
            return new long[] { 1, 0 };
        }
        int startedIdx = started ? 1 : 0;
        if (!tight && memo[pos][startedIdx] != null) {
            return memo[pos][startedIdx];
        }
        int maxDigit = tight ? digits[pos] : 9;
        long completions = 0;
        long occurrences = 0;
        for (int digit = 0; digit <= maxDigit; digit++) {
            long[] inner = solve(digits, d, pos + 1, tight && digit == maxDigit, started || digit > 0, memo);
            completions += inner[0];
            occurrences += inner[1];
            // Placing d here shows d in every completion below, unless it is
            // a leading zero -- those are never written.
            if (digit == d && (started || digit > 0)) {
                occurrences += inner[0];
            }
        }
        long[] state = { completions, occurrences };
        if (!tight) {
            memo[pos][startedIdx] = state;
        }
        return state;
    }
}

File: problems-adapt/1001-1100/1067_digit-occurrences-in-a-range/solution_digit_dp.ts
function countDigitOccurrences(d: number, low: number, high: number): number {
    const countUpTo = (d: number, n: number): number => {
        if (n <= 0) {
            return 0;
        }
        const s = String(n);
        const digits: number[] = [];
        for (let i = 0; i < s.length; i++) {
            digits.push(s.charCodeAt(i) - 48);
        }
        // Free (non-tight) suffixes recur, so they are memoized per
        // (position, started): [completions, occurrences] pairs.
        const memo = new Map<number, [number, number]>();
        // Each state reports how many suffix completions it admits and how
        // many appearances of d those completions contain.
        const solve = (pos: number, tight: boolean, started: boolean): [number, number] => {
            if (pos === digits.length) {
                return [1, 0];
            }
            const key = pos * 2 + (started ? 1 : 0);
            if (!tight && memo.has(key)) {
                return memo.get(key)!;
            }
            const maxDigit = tight ? digits[pos] : 9;
            let completions = 0;
            let occurrences = 0;
            for (let digit = 0; digit <= maxDigit; digit++) {
                const inner = solve(pos + 1, tight && digit === maxDigit, started || digit > 0);
                completions += inner[0];
                occurrences += inner[1];
                // Placing d here shows d in every completion below, unless it
                // is a leading zero -- those are never written.
                if (digit === d && (started || digit > 0)) {
                    occurrences += inner[0];
                }
            }
            const state: [number, number] = [completions, occurrences];
            if (!tight) {
                memo.set(key, state);
            }
            return state;
        };
        // The all-zero completion is the number 0 and carries no
        // appearances, so the walk tallies exactly the integers 1..n.
        return solve(0, true, false)[1];
    };
    // Prefix-count reduction: occurrences in [low, high] = f(high) - f(low-1).
    return countUpTo(d, high) - countUpTo(d, low - 1);
}
